[Kernel] Add more dtype support for GGUF kernels

SzymonOzog opened the pull request:
This adds float and bfloat16 support for GGUF. 

The test comment on bfloat comes from this result(and a couple of simillar ones):

```
    def test_mmq(num_tokens: int, hidden_size: int, dtype: torch.dtype,
                 quant_type: GGMLQuantizationType):
        current_platform.seed_everything(0)

        tensors = get_gguf_sample_tensors(hidden_size, quant_type)
        x = torch.rand((num_tokens, hidden_size), dtype=dtype, device="cuda")
        for tensor in tensors:
            weight = torch.tensor(dequantize(tensor.data, quant_type),
                                  device="cuda").to(dtype)
            ref_output = x @ weight.T

            qweight = torch.tensor(tensor.data, device="cuda")
            output = ops.ggml_mul_mat_a8(qweight, x, quant_type,
                                         qweight.shape[0]).to(dtype)

            ids = torch.isclose(output, ref_output, atol=1, rtol=1e-1).logical_not()
            print(weight.mean())
            print(output[ids])
            print(ref_output[ids])
>           torch.testing.assert_close(output, ref_output, atol=1, rtol=1e-1)
E           AssertionError: Tensor-likes are not close!
E
E           Mismatched elements: 19 / 1572864 (0.0%)
E           Greatest absolute difference: 1.3515625 at index (1217, 6) (up to 1 allowed)
E           Greatest relative difference: 1968.0 at index (329, 628) (up to 0.1 allowed)

tests/kernels/test_gguf.py:133: AssertionError
----- Captured stdout call ------------
tensor(6.0797e-05, device='cuda:0', dtype=torch.bfloat16)
tensor([ 1.0547,  0.9062,  0.9141,  1.2734,  1.1016,  1.1797, -1.1094, -0.4902,
        -1.6797, -0.6094, -1.6797,  0.2656,  1.2344,  2.9375,  0.3750, -1.5234,
        -1.3984,  0.9492,  2.7344], device='cuda:0', dtype=torch.bfloat16)
tensor([ 1.9409e-02, -1.2109e-01, -1.3379e-01,  1.1133e-01,  6.2500e-02,
         1.4062e-01,  5.6458e-04,  5.8594e-01, -5.8203e-01,  5.1562e-01,
        -5.0000e-01, -1.0859e+00,  1.7676e-01,  1.7344e+00, -6.9922e-01,
        -4.7070e-01, -7.0801e-02, -1.2695e-01,  1.5156e+00], device='cuda:0',
       dtype=torch.bfloat16)
 ```

Commit:
Add more dtype support to matvec

Signed-off-by: SzymonOzog <szymon.ozog@aleph-alpha.com>

Commit:
Add more dtype support to matmul

Signed-off-by: SzymonOzog <szymon.ozog@aleph-alpha.com>

Commit:
Add support python scripts

Signed-off-by: SzymonOzog <szymon.ozog@aleph-alpha.com>

Commit:
add tests

Signed-off-by: SzymonOzog <szymon.ozog@aleph-alpha.com>

Commit:
correct dtype for fakes

Signed-off-by: SzymonOzog <szymon.ozog@aleph-alpha.com>

cjackal commented on tests/kernels/test_gguf.py:
```diff
@@ -23,8 +23,8 @@ def get_gguf_sample_tensors(
 
-DTYPES = [torch.half]
-# Hidden_size for testing, must match the sample file in HF repo,
+DTYPES = [torch.half, torch.bfloat16, torch.float32
+          ]  # Hidden_size for testing, must match the sample file in HF repo,
```
small typo here:
```suggestion
DTYPES = [torch.half, torch.bfloat16, torch.float32]
# Hidden_size for testing, must match the sample file in HF repo,
```

SzymonOzog replied:
Must have accidentally joined lines, fixed. Thanks for spotting

cjackal commented:
I haven't validated model output thoroughly but it seems working well, and the CI failure (on auto config resolution) looks irrelevant.

SzymonOzog committed:
fix comment alignment

Signed-off-by: SzymonOzog <szymon.ozog@gmail.com>

cjackal commented:
@SzymonOzog Could we run CI again (e.g. git push with `--allow-empty`)? I think the failing tests are irrelevant, so on the next run this PR would be green and vllm maintainers may get noticed and start review.

Isotr0py reviewed (approved):
Thanks for this awesome enhancement!

Isotr0py commented:
Added ready label to trigger the kernel tests :)

Isotr0py commented:
Please address the failing kernel tests.

Commit:
another bounds check

Signed-off-by: SzymonOzog <szymon.ozog@aleph-alpha.com>

SzymonOzog commented:
@Isotr0py 
Fixed it, it was caused by the matmul kernel reading out of memory when the weights shape wasn't a multiple of 512, I added another boundary check to address this. Wasn't caught initially by running tests because the memory is zeroed out when running tests in isolation so it doesn't affect the output. If you want to reproduce locally it's easiest to just force the memory around the matmul to be nan:

```py
qweight = torch.tensor(tensor.data, device="cuda")
nan_mem = torch.full_like(qweight, 255)
qweight = torch.cat((qweight, nan_mem)).contiguous()
qweight = qweight[:qweight.shape[0]//2]
``` 

Could you rerun the tests?

cjackal commented:
The failing tests in entrypoints test are fixed on main(#14052), FYI
